### PR TITLE
[fast-ut] [reduced-it] Add AnsiCast corner cases [databricks]

### LINKED ISSUE #15621: [AutoSparkUT] Add AnsiCast corner-case IT coverage
## Coverage gap

AnsiCast had no Python integration test combining broad source-type corner coverage, CPU/GPU equality, and proof that the ANSI cast itself executed on GPU without CPU Cast/AnsiCast fallback.

## Test change

Add one parameterized test covering all 15 Spark 3.3 source types currently supported by GPU AnsiCast: Boolean, Byte, Short, Int, Long, Float, Double, Decimal, String, Date, Timestamp, Array, Map, Struct, and DayTimeInterval.

The matrix includes nulls, finite numeric minima/maxima, positive and negative zero, NaN, empty/blank/ASCII/Unicode strings, nested nullability, and interval bounds. Positive and negative infinity are intentionally out of scope for this milestone. The floating-result case compares the sign bit of zero, and every GPU plan requires `GpuProjectExec` with no CPU `AnsiCast` or `Cast` fallback.

YearMonthInterval is intentionally excluded: a Spark 3.3 probe with ConstantFolding disabled showed that casting YearMonthInterval to String falls back through `GpuCpuBridgeExpression`, so it is not valid GPU AnsiCast coverage.

## Local validation evidence

Spark 3.3.0 / Python 3.10, current head `9b22ec747`:

```text
cast_test.py::test_ansicast_corner_cases
15 passed, 34877 deselected, 29 warnings in 14.71s
```

Spark 3.5.0 / shim350, current Python test code:

```text
15 passed, 34883 deselected, 21 warnings in 14.96s
covered lines: +0
```

### PR DESCRIPTION
**Pull request by wjxiz1992**
**JaCoCo sql-plugin line coverage: +0 lines** (shim350 vs nightly b14)

Closes #15621

## Summary

- Add one parameterized Python IT for the 15 Spark 3.3 source types supported by GPU AnsiCast.
- Cover nulls, finite numeric bounds, positive and negative zero, NaN, empty/blank/Unicode strings, complex nested types, and DayTimeInterval bounds.
- Verify CPU/GPU result parity, explicitly compare the sign bit of floating zero, require `GpuProjectExec`, and reject CPU `AnsiCast`/`Cast` fallback in the GPU plan.
- Keep positive and negative infinity outside this milestone, as agreed.

YearMonthInterval is not credited: with ConstantFolding disabled, Spark 3.3 falls back through `GpuCpuBridgeExpression` for its cast, so it is not valid GPU AnsiCast coverage.

## Validation

Spark 3.3.0 / Python 3.10, review-fix commit `60fcd4ded`:

```text
TZ=UTC cast_test.py::test_ansicast_corner_cases
15 passed, 522 deselected, 18 warnings in 10.66s

TZ=Asia/Shanghai cast_test.py::test_ansicast_corner_cases
15 passed, 522 deselected, 2 warnings in 10.66s
```

Spark 3.5.0 / shim350, review-fix test code with nightly b14 classfiles:

```text
cast_test.py::test_ansicast_corner_cases
15 passed, 522 deselected, 18 warnings in 11.62s
covered lines: +0
```

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by wjxiz1992** [comment]
build

File: integration_tests/src/main/python/cast_test.py [hunk 1]
```diff
@@ -76,6 +76,92 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+    def do_cast(spark):
+        return unary_op_df(spark, data_gen).select(
+            f.col('a').cast(to_type).alias('result'))
+
+    (from_cpu, cpu_df), (from_gpu, gpu_df) = run_with_cpu_and_gpu(
+        do_cast, 'COLLECT_WITH_DATAFRAME', conf=ansi_enabled_conf)
+
+    callback = spark_jvm().org.apache.spark.sql.rapids.ExecutionPlanCaptureCallback
+    callback.assertContainsAnsiCast(cpu_df._jdf)
+    callback.assertContainsAnsiCast(gpu_df._jdf)
```
[the remark this answers is not in the document]
Comment (wjxiz1992) [reply integration_tests/src/main/python/cast_test.py]: Thanks — updated in 9b22ec747. The GPU plan now requires GpuProjectExec and explicitly asserts no CPU AnsiCast or Cast fallback. Spark 3.3 / Python 3.10 validation: 15 passed, 0 failed.

File: integration_tests/src/main/python/cast_test.py [hunk 2]
```diff
@@ -76,6 +76,92 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+    assert_equal(from_cpu, from_gpu)
```
[the remark this answers is not in the document]
Comment (wjxiz1992) [reply integration_tests/src/main/python/cast_test.py]: Updated in 9b22ec747. The floating-result case now compares math.copysign for every zero and asserts that negative zero was exercised. Spark 3.3 / Python 3.10 validation: 15 passed, 0 failed.

File: integration_tests/src/main/python/cast_test.py [hunk 3]
```diff
@@ -76,6 +78,110 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+            special_cases=[DOUBLE_MIN, DOUBLE_MAX, 0.0, -0.0, 1.0, -1.0, float('nan')]),
+        BooleanType(),
+        id='double-to-boolean'),
+    pytest.param(DecimalGen(precision=38, scale=10), StringType(), id='decimal-to-string'),
+    pytest.param(
+        StringGen('(| |true|hello|中文|😀)'),
+        BinaryType(),
+        id='string-to-binary'),
+    pytest.param(DateGen(), StringType(), id='date-to-string'),
+    pytest.param(TimestampGen(), StringType(), id='timestamp-to-string'),
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: `TimestampType -> StringType` is not GPU-supported in non-UTC time zones, but this test unconditionally requires a GPU ANSI cast. I reproduced this with `TZ=Asia/Shanghai`: the other 14 cases passed, while this case failed at `assertContainsAnsiCast`. Since the non-UTC nightly runs the full suite, could we change this to `TimestampType -> DateType`, or skip this parameter outside UTC?

File: integration_tests/src/main/python/cast_test.py [hunk 4]
```diff
@@ -12,10 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import math
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: nit: `math` is already imported below; could we keep only one import?

File: integration_tests/src/main/python/cast_test.py (hunk 3 above)
Comment (wjxiz1992) [reply integration_tests/src/main/python/cast_test.py]: Thanks, updated.

File: integration_tests/src/main/python/cast_test.py (hunk 4 above)
Comment (wjxiz1992) [reply integration_tests/src/main/python/cast_test.py]: Updated.

**Comment by wjxiz1992** [comment]
@thirtiseven Could you please review this PR when you have a chance? Thanks.

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
build

File: integration_tests/src/main/python/cast_test.py [hunk 5]
```diff
@@ -76,6 +77,110 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
         conf=ansi_enabled_conf,
         error_message="SparkRuntimeException")
 
+
+# YearMonthIntervalType is excluded because its ANSI casts are not GPU-supported in Spark 3.3.
+@allow_non_gpu(*non_utc_allow)
+@pytest.mark.parametrize('data_gen,to_type', [
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: nit: Could we move this case matrix into a named file-local list, such as `_ansi_cast_corner_case_params`, so the decorator does not obscure the test itself?

File: integration_tests/src/main/python/cast_test.py [hunk 6]
```diff
@@ -76,6 +77,110 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+    def do_cast(spark):
+        return unary_op_df(spark, data_gen).select(
+            f.col('a').cast(to_type).alias('result'))
+
+    (from_cpu, cpu_df), (from_gpu, gpu_df) = run_with_cpu_and_gpu(
+        do_cast, 'COLLECT_WITH_DATAFRAME', conf=ansi_enabled_conf)
+
+    callback = spark_jvm().org.apache.spark.sql.rapids.ExecutionPlanCaptureCallback
+    callback.assertContainsAnsiCast(cpu_df._jdf)
+    callback.assertContainsAnsiCast(gpu_df._jdf)
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: Could we add an `assert_contains_ansi_cast(df)` helper to `asserts.py` and use it here and in `hive_write_test.py`? The Scala callback already provides the shim-aware implementation, so a Python wrapper would keep the JVM/Py4J plumbing out of individual tests without coupling the helper to a CPU/GPU pair.

File: integration_tests/src/main/python/cast_test.py [hunk 7]
```diff
@@ -76,6 +77,110 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+    gpu_plan = gpu_df._jdf.queryExecution().executedPlan()
+    callback.assertContains(gpu_plan, 'GpuProjectExec')
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: Could this use the existing `@validate_execs_in_gpu_plan('GpuProjectExec')` marker? Test mode also rejects CPU expressions, including those inside `GpuCpuBridgeExpression`, so using a narrower non-UTC allowlist that does not permit `Cast` should remove the manual fallback checks.

File: integration_tests/src/main/python/cast_test.py [hunk 8]
```diff
@@ -76,6 +77,110 @@ def test_cast_string_to_boolean_invalid_ansi_on(invalid_value):
...
+        for cpu_row, gpu_row in zip(from_cpu, from_gpu):
+            cpu_value = cpu_row['result']
+            gpu_value = gpu_row['result']
+            if cpu_value is not None and cpu_value == 0.0:
+                cpu_sign = math.copysign(1.0, cpu_value)
+                gpu_sign = math.copysign(1.0, gpu_value)
+                assert cpu_sign == gpu_sign, \
+                    f'GPU changed the sign of zero: CPU={cpu_value}, GPU={gpu_value}'
+                saw_negative_zero = saw_negative_zero or cpu_sign < 0
+        assert saw_negative_zero, 'Float corner data did not exercise negative zero'
```
Comment (thirtiseven) [inline integration_tests/src/main/python/cast_test.py]: Could we make the negative-zero input deterministic, for example with `RepeatSeqGen` or a dedicated case? `DataGen.special_cases` are sampled by weight, so `saw_negative_zero` can fail for a seed that does not select -0.0. The signed-zero comparison itself could then use an opt-in recursive helper in `asserts.py`.

**Comment by wjxiz1992** [comment]
build

**Review by thirtiseven — APPROVED** [review]
LGTM
